vernac/lemmas.mli

```diff
@@ -13,23 +13,34 @@ open Decl_kinds
 
 type declaration_hook
 
-val mk_hook : (Decl_kinds.locality -> GlobRef.t -> unit) -> declaration_hook
-val call_hook :
-  ?hook:declaration_hook -> ?fix_exn:Future.fix_exn ->
-  Decl_kinds.locality -> GlobRef.t -> unit
+(* Hooks allow users of the API to perform arbitrary actions at
+ * proof/definition saving time. For example, to register a constant
+ * as a Coercion, perform some cleanup, update the search database,
+ * etc...
+ *
+ * Here, we use an extended hook type suitable for obligations /
+ * equations.
+ *)
+(** [hook_type] passes to the client:
+    - [ustate]: universe constraints obtained when the term was closed
+    - [(n1,t1),...(nm,tm)]: XXXX
+    - [locality]: Locality of the original declaration
+    - [ref]: identifier of the origianl declaration
```

Comment:
[(n1, t1)] .. association list between obligation name and the corresponding defined term (might be a constant, but also an arbitrary term in the Expand case of obligations)
[ref] global reference of the declared definition

The first mapping could be generalized to a mapping between goal id (however we chose to encode it) and the defined constants corresponding to it (e.g. if Qed were to define all the initial goals as constants for example)

Reply:
Thanks @mattam82 , updating. Indeed I think this can be made general-purpose.